Added play media to squeezebox

sytone opened the pull request:
**Description:**
Updated the squeezebox component to handle play commands. 
- You can now specify a URI in the play and it will play it immediately. 
- You can specify enqueuer and it adds it to the end of the current play list. 

**Related issue (if applicable):** NA

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#
NA - adding support for existing features, no change in documentation.

**Example entry for `configuration.yaml` (if applicable):**
NA - No configuration changes. 

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
- [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If code communicates with devices, web services, or a:
- [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16)).
- [ ] New dependencies are only imported inside functions that use them ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51)).
- [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
- [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
- [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [ ] Tests have been added to verify that the new code works.

The squeezebox component can now add a URI to an existing playlist or just over write it to force a stream to play.

sytone committed:
Added play media to squeezebox

The squeezebox component can now add a URI to an existing playlist or just over write it to force a stream to play.

sytone committed:
Cleaned up flake8 issues with formatting. 

Spacing... The end of the world! Fixed. Once day the tools will fix this on the fly, one day...

[x] ./homeassistant/components/media_player/squeezebox.py:307:1: W293 blank line contains whitespace
[x] ./homeassistant/components/media_player/squeezebox.py:366:1: W391 blank line at end of file
[x] ./homeassistant/components/media_player/squeezebox.py:366:1: W293 blank line contains whitespace

Updated SUPPORT_SQUEEZEBOX to add SUPPORT_PLAY_MEDIA

[x] ./homeassistant/components/media_player/squeezebox.py:13:1: F401 'homeassistant.components.media_player.SUPPORT_PLAY_MEDIA' imported but unused

kellerza commented on homeassistant/components/media_player/squeezebox.py:
```diff
@@ -303,3 +304,62 @@ def turn_on(self):
+
+        <playerid> playlist play <item> <title> <fadeInSecs>
+
+        The "playlist play" command puts the specified song URL,
```
Can you please move some of this documentation to a PR against homeassistant.io? Lets try to keep this short and to the point

sytone replied:
This documentation is not for the user, it is for any future developer. I has always tried to add as much information in the code so future developers can work better with it. Is this not a stance for Home Assistant?

kellerza commented on homeassistant/components/media_player/squeezebox.py:
```diff
@@ -303,3 +304,62 @@ def turn_on(self):
+        if kwargs.get(ATTR_MEDIA_ENQUEUE):
+            self.add_uri_to_playlist(media_id)
+        else:
+            self.play_uri(media_id)
```
add_uri_to_playlist and play_uri is private functions, can you rename them to start with an underscore? `_play_uri`

Even better would be to remove them completely and just determine the `play` or `add` parameter

```python
enque = 'add' if kwargs.get(ATTR_MEDIA_ENQUEUE) else 'play'
self._lms.query(self._id, 'playlist', enque, media_id)
```

And then move all docs to a homeassistant.io PR

sytone replied:
This may be a coding style conflict. I am never a fan of overloading functions or methods, yes I can collapse them down but I was hoping to update the fade time as a property in a future change which would complicate the if statement more. From a python execution point is the separate function causing a decrease in performance? I am not deeply engaged with the python platform and do not know the full performance characteristics but typically having additional calls like this are non-impactful on performance. 

Let me know where coding style and preferences are for this project so I have better visibility in the future. I have updated with the underscore (_) to indicate they are private.

sytone committed:
Updates from review

Updated the comments to indicate they are developer / API comments and not for end users.
Marked the private functions with a leading underscore (_)

sytone committed:
Fixed Lint issues. 

202ERROR: InvocationError: '/home/travis/build/home-assistant/home-assistant/.tox/lint/bin/flake8'

203lint runtests: commands[1] | pylint homeassistant

204************* Module homeassistant.components.media_player.squeezebox

205C:322, 0: Trailing whitespace (trailing-whitespace)

kellerza commented:
Let's try to keep comments as short as possible, only the core HASS methods are documented in detail. How play and enque work fits well in user documentation. You can keep the telnet command sructures showing additional features (Like fadeinsec). Examples and returns feels a bit redundant.

"Overloading" methods really depends on the component and services it exposes (see [here](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/media_player/__init__.py#L389) for the play service)). Not all media players will support all functions and we have to try model them in a way (part of the reason I asked to make new functions private/even remove them). You are welcome to suggest enhancements to the media player model, but preferably in a separate PR.

Phigital-Inc commented:
Hi, to support media_content_type: playlist, I would recommend not just using the "playlist play" CLI command but rather then "playlistcontrol" CLI command -- the former does not support playing imported iTunes playlists nor specifying "playlist_name". Ideally, the "playlistcontrol" command would support URIs, but unfortunately it does not...

Phigital-Inc commented:
CORRECTION: Turns out the "playlistcontrol" does support URIs (albeit a little slowly since it has to load the playlist), so it could replace "playlist play" as the default
